documentation/developer/adr/0004-record-parameter-frame.md

```diff
@@ -0,0 +1,89 @@
+# 4. Record ParameterFrame decisions
+
+Date: 2022-06-30
+Updated: 2022-10-24
+
+## Status
+
+Completed
+
+## Context
+
+We want to make it easier to configure a `ParameterFrame` for a `ReactorDesigner`.
+
+Currently we have all `Parameters` listed in `Configuration`. There are a number of downsides to this approach.
+
+* Their definition is global meaning that no shared names can exist across different `Builder`s.
+* `Builder`s take more `Parameters` than they probably need
+
+Options:
+
+* Define a `ParameterFrame` for each `Builder` with no default values.
+  * Reconstruct the global list of required `Parameters` for a `ReactorDesigner` from
+    the set of `Parameters` required for all of its `Builder`s.
+* Nested / indexed / tagged `ParameterFrame` where `Builder`s would automatically tag
+  a `Parameter` but the overall list would remain flat.
+  * A `ConfigurationSchema`-like approach for each `Builder`.
+
+e.g.
+
+```python
+class BuilderOne(Builder):
+    R_0: Parameter[float]
+    A: Parameter[float]
+```
+
+## Decision
+
+We will trial the three options above  taking e.g. the EUDEMO `PlasmaBuilder` as an example.
+
+#### 2022-10-24
+
+The final design landed on consists of some of both options:
+
+-  A dedicated `ParameterFrame` is created for each object that uses one
+
+- No default values are stored on a `ParameterFrame`
+
+  - A `MappedParameterFrame` has defaults and is used exclusively for external codes where some variables are not controlled in mapping but can be updated through the solver interface. Eg. PROCESS DEMO style input
+
+-  Only the value and source of a `Parameter` have an API entry point to be updated after creation
+
+- A `ParameterFrame` definition looks like one of the below. The decorated version is a shortcut for the dataclass version but there is no static autocomplete as of writing. Therefore within bluemira core code the dataclass version should be used.
+
+  ```python
+  from dataclasses import dataclass
+  from bluemira.base.parameter_frame import (
+      Parameter,
+      ParameterFrame,
+  )
+
+  @dataclass
+  class MyFrameForAJob(ParmeterFrame):
+     R_0: Parameter[float]
+     A: Parameter[float]
+
+  ```
+
+A decorator to allow for easier definition of a `ParameterFrame` was also developed but typing issues need to be fixed to enable its use:
+
+  ```python
+  from bluemira.base.parameter_frame import (
+      Parameter,
+      ParameterFrame,
+      parameter_frame,
+  )
+
+  @parameter_frame
+  class MyDecoratedFrameForAJob:
+      B: Parameter[float]
+      z_0: Parameter[float]
+  ```
+
+Further work will be done to add helper methods to collate all `Parameters` required in a given piece of code
+
+## Consequences
+
+All existing metadata has been removed from bluemira and should be stored in a `Parameter` JSON file for a given reactor design.
+
+All `Parameters` need to be defined for a given design.
```

Comment:
- We have many different `ParameterFrame` classes instead of one global one.
- Reactor Designers are expected to define their own parameter frames.

I'm sure there are more, but I'm struggling to come up with any on a Friday afternoon